File: 400/491 - Increasing Subsequences.java
import java.util.ArrayList;
import java.util.HashSet;

class Solution {
    public List<List<Integer>> findSubsequences(int[] nums) {
        HashSet<List<Integer>> set = new HashSet<List<Integer>>();
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        
        helper(nums, 0, new ArrayList<Integer>(), set);
        
        ret.addAll(set);
        return ret;
    }
    
    private void helper(int[] nums, int from, List<Integer> prefix, HashSet<List<Integer>> set) {
        int i;

        if (prefix.size() >= 2)
            set.add(new ArrayList<Integer>(prefix));
        
        for (i = from; i < nums.length; ++i) {
            if ((! prefix.isEmpty()) && prefix.get(prefix.size() - 1) > nums[i])
                continue;
            prefix.add(nums[i]);
            if (! set.contains(prefix))
                helper(nums, i + 1, prefix, set);
            prefix.remove(prefix.size() - 1);
        }
    }
}

File: csharp/491.increasing-subsequences.cs
/*
 * @lc app=leetcode id=491 lang=csharp
 *
 * [491] Increasing Subsequences
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private void FindSubsequencesRecurse(int[] nums, IList<IList<int>> ret, IList<int> curr, int i) {
        if (curr.Count >= 2) {
            ret.Add(new List<int>(curr));
        }
        ISet<int> visited = new HashSet<int>();
        for (int j = i; j < nums.Length; ++j) {
            int num = nums[j];
            if (visited.Contains(num) || (curr.Count > 0 && num < curr[curr.Count - 1])) {
                continue;
            }
            visited.Add(num);
            curr.Add(num);
            FindSubsequencesRecurse(nums, ret, curr, j + 1);
            curr.RemoveAt(curr.Count - 1);
        }
    }

    public IList<IList<int>> FindSubsequences(int[] nums) {
        var ret = new List<IList<int>>();
        FindSubsequencesRecurse(nums, ret, new List<int>(), 0);
        return ret;
    }
}
// @lc code=end

